python_ta/checkers/possibly_undefined_checker.py

```diff
@@ -0,0 +1,157 @@
+        if self._check_possibly_undefined(node):
+            self.add_message('possibly-undefined', node=node)
+
+    def visit_module(self, node):
+        self._analyze(node)
+
+    def visit_functiondef(self, node):
+        self._analyze(node)
+
+    def _check_possibly_undefined(self, node: astroid.Name) -> bool:
+        """Returns True if there exists a path from the start block to node where
+        a variable used by node might not be defined."""
+        return node in self.possibly_undefined
+
+    def _analyze(self, node: Union[astroid.Module, astroid.FunctionDef]) -> None:
+        """Runs the data flow algorithm on a `Module` or `Function` CFG, which in turn
+        appends `Name` nodes to `possibly_undefined` if it might not be defined.
+        """
+        facts = {}
+        blocks = self._get_blocks_po(node)
+
+        all_assigns = self._get_assigns(blocks)
+        for block in blocks:
+            facts[block] = {}
+            facts[block]['out'] = all_assigns.copy()
+
+        worklist = blocks[:]
+        while len(worklist) != 0:
+            b = worklist.pop()
+            outs = [facts[p.source]['out'] for p in b.predecessors if p.source in facts]
+            if outs == []:
+                facts[b]['in'] = set()
+            else:
+                facts[b]['in'] = set.intersection(*outs)
+            temp = self._transfer(b, facts[b]['in'], all_assigns)
+            if temp != facts[b]['out']:
+                facts[b]['out'] = temp
+                successors = set([succ.target for succ in b.successors])
+                worklist = list(set(worklist).union(successors))
+
+    def _transfer(self, block: CFGBlock, in_facts: Set[str], all_assigns: Set[str]) -> Set[str]:
+        gen = set()
+        kill = set()
+        for statement in block.statements:
+            if isinstance(statement, astroid.Assign):
+                gen.update(self._get_targets(statement))
+            elif isinstance(statement, astroid.AnnAssign) and hasattr(statement.target, 'name'):
+                gen.add(statement.target)
+            elif isinstance(statement, astroid.Delete):
+                kill.update(self._get_targets(statement))
+            self._analyze_statement(statement, gen.union(in_facts.difference(kill)), all_assigns)
+        return gen.union(in_facts.difference(kill))
+
+    def _analyze_statement(self, node: NodeNG, facts: Set[str], local_vars: Set[str]) -> None:
+        if isinstance(node, astroid.Name):
+            name = node.name
+            if name in local_vars and name not in facts and not self._is_function_name(node):
+                self.possibly_undefined.add(node)
+            elif node in self.possibly_undefined:
+                self.possibly_undefined.remove(node)
+        else:
+            for child in node.get_children():
+                self._analyze_statement(child, facts, local_vars)
+
+    def _get_targets(self, statement: Union[astroid.Assign, astroid.Delete]) -> Set[str]:
```

Comment:
There's probably a more natural astroid way of doing this. At the very least, using `nodes_of_class(astroid.AssignName)`, but could be even better.